File: old-cpp-lib/xfs_extent.cpp
#include "xfs_extent.hpp"

#include <endian.h> // be64toh

using uf::xfs::Extent;

Extent::Extent(
    xfs_fileoff_t FileOffset,
    xfs_fsblock_t BlockOffset,
    xfs_filblks_t BlockCount,
    xfs_exntst_t  State) noexcept
    : m_fileOffset(FileOffset)
    , m_blockCount(BlockCount)
    , m_blockOffset(BlockOffset)
    , m_state(State)
{}

Extent::Extent(xfs_bmbt_rec_t DiskExtent) noexcept
{
    // the extent is stored as two big-endian 64 bit chunks
    DiskExtent.l0 = be64toh(DiskExtent.l0);
    DiskExtent.l1 = be64toh(DiskExtent.l1);

    /*
     * Bmap btree record and extent descriptor.
     *  l0:63           : extent flag (value 1 indicates non-normal).
     *  l0:9-62         : file offset.
     *  l0:0-8 l1:21-63 : block offset.
     *  l1:0-20         : block count.
     */
    m_state      = static_cast<xfs_exntst_t>(DiskExtent.l0 & (~(static_cast<uint64_t>(-1ULL) >> BMBT_EXNTFLAG_BITLEN)));
    m_fileOffset = DiskExtent.l0 & BMBT_STARTOFF_MASK;
    m_blockOffset = ((DiskExtent.l0 & ~BMBT_STARTOFF_MASK) << (63ULL - BMBT_STARTBLOCK_BITLEN)) +
                    ((DiskExtent.l1 & ~BMBT_BLOCKCOUNT_MASK) >> BMBT_BLOCKCOUNT_BITLEN);
    m_blockCount = DiskExtent.l1 & BMBT_BLOCKCOUNT_MASK;
}

bool Extent::IsValid(const xfs_superblock_t& Superblock) const noexcept
{
    // extent has been preallocated but has not yet been written
    if (XFS_EXT_UNWRITTEN == m_state)
    {
        return false;
    }

    // zero length extents are not worth considering
    if (0 == m_blockCount)
    {
        return false;
    }

    // ignore unused extents (zeroed out)
    // could remove br_state from this check because of the previous check
    if (0 == m_blockCount && 0 == m_blockOffset && 0 == m_fileOffset && 0 == m_state)
    {
        return false;
    }

    // ignore extents beyond the filesystem
    // could there be an overflow here?
    if (GetInputOffset(Superblock) + m_blockCount > be64toh(Superblock.sb_dblocks))
    {
        return false;
    }

    return true;
}

uint64_t Extent::GetInputOffset(const xfs_superblock_t& Superblock) const noexcept
{
    // br_startblock is an unsigned 64 bit filesystem block number combining AG number and block offset into the AG.

    auto agIndex = static_cast<xfs_agnumber_t>(
        (m_blockOffset & ~((static_cast<uint64_t>(-1)) << Superblock.sb_agblklog)) >> Superblock.sb_agblklog);

    xfs_fsblock_t agRelativeOffset = m_blockOffset & ~((static_cast<uint64_t>(-1)) << Superblock.sb_agblklog);
    xfs_filblks_t blocksPerAG      = be32toh(Superblock.sb_blocksize);

    return agIndex * blocksPerAG + agRelativeOffset;
}

File: src/xfs_extent.zig
const std = @import("std");

pub const inode_entry = @import("inode_entry.zig").inode_entry;
pub const xfs_inode_t = @import("xfs_inode.zig").xfs_inode_t;
pub const xfs_error = @import("xfs_error.zig").xfs_error;
pub const xfs_superblock = @import("xfs_superblock.zig").xfs_superblock;

const c = @import("c.zig").c;
const xfs_mask64lo = @import("c.zig").xfs_mask64lo;
const libxfs_bmbt_disk_get_all = @import("c.zig").libxfs_bmbt_disk_get_all;

pub const xfs_extent_t = struct {
    block_offset: c.xfs_fsblock_t,
    block_count: c.xfs_filblks_t,
    file_offset: c.xfs_fileoff_t,
    state: c.xfs_exntst_t,

    pub fn create(packed_extent: *const c.xfs_bmbt_rec_t) xfs_extent_t {
        var irec: c.xfs_bmbt_irec = undefined;

        libxfs_bmbt_disk_get_all(packed_extent, &irec);

        return xfs_extent_t{
            .block_count = irec.br_blockcount,
            .block_offset = irec.br_startblock,
            .file_offset = irec.br_startoff,
            .state = irec.br_state,
        };
    }

    pub fn check(self: *const xfs_extent_t, superblock: *const xfs_superblock) xfs_error!void {
        if (c.XFS_EXT_UNWRITTEN == self.state) {
            return xfs_error.xfs_ext_unwritten;
        }
        if (0 == self.block_count and 0 == self.block_offset and 0 == self.file_offset and 0 == self.state) {
            return xfs_error.xfs_ext_zeroed;
        }
        if (self.get_input_offset(superblock) + self.block_count > superblock.sb_dblocks) {
            return xfs_error.xfs_ext_beyond_sb;
        }
    }

    pub fn get_input_offset(self: *const xfs_extent_t, superblock: *const xfs_superblock) u64 {
        const agblklog = @as(u64, std.math.maxInt(u64)) << @intCast(superblock.sb_agblklog);

        const ag_relative_offset: c.xfs_fsblock_t = self.block_offset & ~agblklog;

        const ag_index: c.xfs_agnumber_t = @intCast(ag_relative_offset >> @intCast(superblock.sb_agblklog));

        const blocks_per_ag: c.xfs_filblks_t = c.be32toh(superblock.sb_blocksize);

        return ag_index * blocks_per_ag + ag_relative_offset;
    }
};
